### ISSUE TITLE
as-is source code is pretty nonfunctional on Linux

### ISSUE THREAD

**Issue by hsartoris** [alexgooding/auto-drummer#1]
Let me first say: I really like this project, and I plan on using it in my production.

The gist of this issue is to suggest that you just **remove the line on the Releases page directing non-Windows users to run from source.**

I'm running Arch Linux x86_64. In order to get the code running natively, I modified every path to exchange '\\\\' for '/', as well as removed the `startupinfo` variable, which, as you may have already seen on the reddit thread, is not supported on non-Windows systems. 

There may have been another couple of modifications I made, but they aren't all that important given that the program makes a call to clingo.exe, requiring that the user have a working Wine installation already. Given this, and that Auto Drummer.exe runs fine in Wine, I don't think there's much point in trying to run natively.

**Comment by alexgooding** [comment]
That’s great! I’m glad to hear you like the project.

Thanks for brining this to my attention. To be honest, I completely forgot that the solver was an executable file. Another user on macOS has had the same problem. I will amend the instructions.

I’m very unfamiliar with Linux - is Wine easy to use? Is it the kind of software that most people on Linux would already own and know how to use?




________________________________

**Comment by hsartoris** [comment]
If everything is working as intended, it's very easy to use. In this case you would run `wine 'Auto Drummer.exe'`.

Wine is free, and I'd wager that every serious Linux user will have at least a little experience with it, as it's probably the most widely used compatibility layer for running Windows executables on Linux.

The biggest issue people usually run into is discovering that the executable they're running requires some DLLs not provided with the base install of Wine for their distro. This usually results in an error window informing the user that this or that DLL is missing, and googling that message will most often inform the user what additional packages they need to install through Wine, the process for which is a little opaque at first but frankly pretty par for the course for troubleshooting on Linux.

Your executables run fine for me without any fiddling around (although I didn't try with fluidsynth installed), but my Wine is a beta version equipped with a lot of stuff for running FL Studio, so that may not be indicative of general results.

Next time I set up a new installation I'll try running it with a bone stock Wine install and, if anything crops up, I'll start an issue with the process for getting the required dependencies. 

TL;DR: shouldn't be a big deal for most Linux users

**Comment by hsartoris** [comment]
Separately, since clingo is cross-platform, it should be possible to refactor the Auto Drummer source to be properly cross-platform as well. I wouldn't expect you to do this, but if I have some spare time to do so, I could submit a pull request with the changes. Of course they'd be written so as to not change the execution flow of the program on Windows.

If you'd be interested/willing to tolerate that, let me know.

**Comment by alexgooding** [comment]
That would be great! I’d like to have the software available on all major platforms. However, I may make some fairly drastic changes in the near future. I’m looking at bypassing the clingo.exe solver all together – I should be able to import it as a Python module but it may mean quite an overhaul of the code in composer.py. I’d like the user to have a bit more control over the drum fills and the user input. I don’t think it’s completely clear how the user input tab works. I’ll probably have two separate modes: One for filling in a partially created pattern following the rules in the advanced tab and one that will extend any pattern inputted to a specified length. I’d also like to change the way that fills are created and the way that one bar patterns are extended but that hopefully shouldn’t impact anything related to refactoring code.

 

If you think the changes made when refactoring are not so great that it would be easy for me to continue working on it in the future, that would be great, and please submit a pull request. Otherwise I will hopefully be able to make the changes I listed above in about a month after I’ve finished my writeup for the current version of the software.

**Comment by alexgooding** [comment]
@hsartoris, take a look at @jholmes latest commit. I imagine a lot of the changes necessary have been by him. Hopefully, it's just the Cingo Linux binaries that have to be added.

**Comment by hsartoris** [comment]
Looks like it! The path fixes are probably the majority of the issues.

I'll keep an eye on the project and submit pull requests for additional compatibility if needed.

Thanks for your work!

**Comment by alexgooding** [comment]
No problem! Thanks! I'm pleasantly surprised at the amount of interest in this personal project.
